### PR TITLE
Multiple threads to perform BPE encoding

### PR DESCRIPTION
**Pull request by xihui-wu**

### CODE REVIEW

File: Datasets/TextUnsupervised/TextUnsupervised.swift [hunk 1]
```diff
@@ -185,7 +185,16 @@ public struct TextUnsupervised {
             let path = directory.appendingPathComponent("\(variantDetails.filename)/\(name).csv")
             let documentsFull = try readCSV(in: path)
             let documents = Array(documentsFull[0..<min(documentCount, documentsFull.count)])
-            encodedDocs = documents.map { embedding(for: $0, bpe: bpe) }
+
+            encodedDocs = Array(repeating: [], count: documents.count)
+            DispatchQueue.concurrentPerform(iterations: 10) {
```
Comment (texasmichelle) [inline Datasets/TextUnsupervised/TextUnsupervised.swift:190]: This is neat! Could we do this with encoded files as well?

File: Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift [hunk 2]
```diff
@@ -74,7 +74,7 @@ final class TextUnsupervisedTests: XCTestCase {
                 XCTAssertEqual(example.second.shape[0], 1024)
                 totalCount += 1
             }
-            XCTAssertEqual(totalCount, 12)
+            XCTAssertEqual(totalCount, 64)
```
Comment (texasmichelle) [inline Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift:77]: Will this increase the length of time it takes to run these tests during every presubmit? If so, let's keep the lower document count by being explicit during init.

File: Datasets/TextUnsupervised/TextUnsupervised.swift (hunk 1 above)
Comment (texasmichelle) [inline Datasets/TextUnsupervised/TextUnsupervised.swift:190]: Where does the number 10 come from? Consider making this relative to `documentCount`.

Comment (xihui-wu) [reply Datasets/TextUnsupervised/TextUnsupervised.swift]: Made a concurrentMap extension, please see latest change.

File: Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift (hunk 2 above)
Comment (xihui-wu) [reply Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift]: I increased it to show that with this concurrency a larger number of documents won't increase test time. I'm fine to revert it back to 4.

File: Datasets/TextUnsupervised/TextUnsupervised.swift (hunk 1 above)
Comment (xihui-wu) [reply Datasets/TextUnsupervised/TextUnsupervised.swift]: I just pick it up as a reasonable thread count. In latest implementation, I make thread count to be min(10, documentCount), but I'm all ears to any better suggestion. BTW for WikiText103 documentCount can be up to 30k.

File: Datasets/TextUnsupervised/TextUnsupervised.swift [hunk 3]
```diff
@@ -17,6 +17,23 @@ import Foundation
 import ModelSupport
 import TensorFlow
 
+extension Array {
```
Comment (texasmichelle) [inline Datasets/TextUnsupervised/TextUnsupervised.swift:20]: let's make this `fileprivate` and move to the bottom of the file.

File: Datasets/TextUnsupervised/TextUnsupervised.swift [hunk 4]
```diff
@@ -174,14 +191,11 @@ public struct TextUnsupervised {
             let path = directory.appendingPathComponent("\(variantDetails.filename)/\(name).csv")
             let documentsFull = try readCSV(in: path)
             let documents = Array(documentsFull[0..<min(documentCount, documentsFull.count)])
-            encodedDocs = documents.map { embedding(for: $0, bpe: bpe) }
-        } else {
+            encodedDocs = documents.concurrentMap { embedding(for: $0, bpe: bpe) }
+	    } else {
```
Comment (texasmichelle) [inline Datasets/TextUnsupervised/TextUnsupervised.swift:195]: replace tab with spaces

File: Datasets/TextUnsupervised/TextUnsupervised.swift [hunk 5]
```diff
@@ -174,14 +191,11 @@ public struct TextUnsupervised {
...
             let pathPrefix = directory.appendingPathComponent("\(variantDetails.encodedFileName!)/\(name)").path
-            for i in 0..<documentCount {
-                encodedDocs += [
-                  try readEncoded(in: URL(fileURLWithPath: "\(pathPrefix)/doc_\(i).txt"))
-                ]
-            }
+            encodedDocs = (0..<documentCount).map {URL(fileURLWithPath: "\(pathPrefix)/doc_\($0).txt")}
```
Comment (texasmichelle) [inline Datasets/TextUnsupervised/TextUnsupervised.swift:197]: `swift-format`-approved spacing:
```
encodedDocs = (0..<documentCount).map { URL(fileURLWithPath: "\(pathPrefix)/doc_\($0).txt") }
    .concurrentMap { try! readEncoded(in: $0) }
```

File: Datasets/TextUnsupervised/TextUnsupervised.swift (hunk 3 above)
Comment (xihui-wu) [reply Datasets/TextUnsupervised/TextUnsupervised.swift]: good to know!

File: Datasets/TextUnsupervised/TextUnsupervised.swift (hunk 5 above)
Comment (xihui-wu) [reply Datasets/TextUnsupervised/TextUnsupervised.swift]: I ran swift-format --configuration .swift-format Datasets/TextUnsupervised/TextUnsupervised.swift -I
for entire file

File: Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift (hunk 2 above)
Comment (texasmichelle) [reply Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift]: What happened to these changes? I believe they will resolve the failure in #526

Comment (xihui-wu) [reply Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift]: What's the failure? I didn't touch the test eventually.

Comment (texasmichelle) [reply Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift]: In presubmits, I'm seeing:
```
/swift-models/Tests/DatasetsTests/TextUnsupervised/TextUnsupervisedTests.swift:77: error: TextUnsupervisedTests.testCreateWikiText2WithBpe : XCTAssertEqual failed: ("64") is not equal to ("12") -
```
The test passes for me locally with 12, so I don't quite understand this. I wonder if kokoro is picking up a cached image somehow... 🤔 I'll try running again.

**Pull request merged**